solidity/test/KeepFactorySelectionTest.js

```diff
@@ -0,0 +1,243 @@
+    await restoreSnapshot()
+  })
+
+  describe("initialize", async () => {
+    it("can be called only one time", async () => {
+      // already initialized in before
+      await expectRevert(
+        keepFactorySelection.initialize(keepStakeFactory.address),
+        "Already initialized",
+      )
+    })
+  })
+
+  describe("selectFactory", async () => {
+    // No ETH stake factory set,
+    // No selection strategy set.
+    it("returns KEEP stake factory by default", async () => {
+      const selected = await keepFactorySelection.selectFactory()
+      expect(selected, "unexpected factory selected").to.equal(
+        keepStakeFactory.address,
+      )
+    })
+
+    // ETH stake factory set.
+    // Selection strategy set.
+    it("returns the same factory until refreshed", async () => {
+      await keepFactorySelection.setFullyBackedKeepFactory(
+        ethStakeFactory.address,
+      )
+      await keepFactorySelection.setKeepFactorySelector(
+        keepFactorySelector.address,
+      )
+
+      await keepFactorySelector.setDefaultMode()
+
+      let selected = await keepFactorySelection.selectFactory()
+      expect(selected, "unexpected factory selected").to.equal(
+        keepStakeFactory.address,
+      )
+
+      await keepFactorySelector.setFullyBackedMode()
+
+      // although the strategy wants to return fully-backed factory
+      // selectFactory() is still returning the default option, until
+      // the selection will be refreshed
+      selected = await keepFactorySelection.selectFactory()
+      expect(selected, "unexpected factory selected").to.equal(
+        keepStakeFactory.address,
+      )
+
+      // refresh the choice
+      await keepFactorySelection.selectFactoryAndRefresh()
+
+      // the factory has changed because selectFactoryAndRefresh
+      // refreshed the selection
+      selected = await keepFactorySelection.selectFactory()
+      expect(selected, "unexpected factory selected").to.equal(
+        ethStakeFactory.address,
+      )
+    })
+  })
+
+  describe("selectFactoryAndRefresh", async () => {
+    // No ETH stake factory set.
+    // No selection strategy set.
+    it("returns KEEP stake factory by default", async () => {
+      const selected = await keepFactorySelection.selectFactoryAndRefresh.call()
+      await keepFactorySelection.selectFactoryAndRefresh()
+      expect(selected, "unexpected factory selected").to.equal(
+        keepStakeFactory.address,
+      )
+    })
+
+    // ETH stake factory set.
+    // No selection strategy set.
+    it("returns KEEP stake factory if strategy is not set", async () => {
+      await keepFactorySelection.setFullyBackedKeepFactory(
+        ethStakeFactory.address,
+      )
+      const selected = await keepFactorySelection.selectFactoryAndRefresh.call()
+      await keepFactorySelection.selectFactoryAndRefresh()
+      expect(selected, "unexpected factory selected").to.equal(
+        keepStakeFactory.address,
+      )
```

Comment:
We could add another test here to check that the selected factory stays the same, as this one is from initialization.